File: ui/visual-component/src/main/java/org/visual/component/widget/FpsWidget.java
package org.visual.component.widget;

import javafx.animation.AnimationTimer;
import javafx.scene.control.Label;
import javafx.scene.paint.Color;
import org.visual.component.util.FxUtil;

public class FpsWidget extends Label {
  private final long[] frameTimes = new long[1000];
  private int frameTimeIndex = 0;
  private boolean arrayFilled = false;

  private final Thread t =
      new Thread(
          () -> {
            AnimationTimer frameRateMeter =
                new AnimationTimer() {
                  @Override
                  public void handle(long now) {
                    long oldFrameTime = frameTimes[frameTimeIndex];
                    frameTimes[frameTimeIndex] = now;
                    frameTimeIndex = (frameTimeIndex + 1) % frameTimes.length;
                    if (frameTimeIndex == 0) {
                      arrayFilled = true;
                    }
                    if (arrayFilled) {
                      long elapsedNanos = now - oldFrameTime;
                      long elapsedNanosPerFrame = elapsedNanos / frameTimes.length;
                      double frameRate = 1000000000.0 / elapsedNanosPerFrame;

                      FxUtil.runOnFx(
                          () -> setText(String.format("Current frame rate: %.3f", frameRate)));
                    }
                  }
                };
            frameRateMeter.start();
          });

  public FpsWidget() {
    setTextFill(Color.WHITE);
    toFront();
    t.start();
  }
}

File: ui/visual-model-component/src/main/kotlin/org/visual/model/component/widget/FpsWidget.kt
package org.visual.model.component.widget

import javafx.animation.AnimationTimer
import javafx.scene.control.Label
import javafx.scene.paint.Color

class FpsWidget : Label() {
    private val frameTimes = LongArray(100)
    private var frameTimeIndex = 0
    private var arrayFilled = false

    private val frameRateMeter: AnimationTimer =
        object : AnimationTimer() {
            override fun handle(now: Long) {
                val oldFrameTime = frameTimes[frameTimeIndex]
                frameTimes[frameTimeIndex] = now
                frameTimeIndex = (frameTimeIndex + 1) % frameTimes.size
                if (frameTimeIndex == 0) {
                    arrayFilled = true
                }
                if (arrayFilled) {
                    val elapsedNanos = now - oldFrameTime
                    val elapsedNanosPerFrame = elapsedNanos / frameTimes.size
                    val frameRate = 1000000000.0 / elapsedNanosPerFrame
                    text = String.format("Current frame rate: %.3f", frameRate)
                }
            }
        }

    init {
        textFill = Color.BLACK
        toFront()
        frameRateMeter.start()
    }

    fun stop() {
        frameRateMeter.stop()
    }
}
